### ISSUE TITLE
Attribute arguments should be expressions

### ISSUE THREAD

**Issue by kyouko-taiga** [val-lang/val#76]
Attribute arguments should be parsed as expressions, so that we can write e.g.:
```val
@align(MemoryLayout<Int64>.alignment())
public let n: Int16
```

**Comment by kyouko-taiga** [comment]
Interesting idea. I see no reason why attributes couldn't have labels, but do we have use cases in mind to justify supporting this complexity already? I do not expect attributes to be user-definable in the foreseeable future and the attributes we have now do not need labels, I think.

**Comment by peter-evans** [comment]
To start with I decided to try implementing this with labels for attributes.

I have a very rough implementation which I think is working, but I ran into an interesting situation where the following test is now failing.
https://github.com/hylo-lang/hylo/blob/fc02f38c10b37bfb6f398f8c135608cca5c7878d/Tests/HyloTests/ParserTests.swift#L480-L489

I think the parser is getting confused and thinks that what comes after `@autoclosure` is the attribute's arguments, and it tries to treat `[E]() -> Int)` as an expression. Not sure how to fix it yet. 🤔

**Comment by kyouko-taiga** [comment]
Normally attributes should apply on declarations, so you could try to parse `@autoclosure` as a special case in the parameter annotation parser.

**Comment by kyouko-taiga** [comment]
You can parse `@autoclosure` as a special symbol in the beginning of a parameter type annotation without using `parseAttribute`. As I said, this parser is meant to parse declarations attributes and should probably be renamed `parseDeclAttribute`. `@autoclosure` is _not_ a declaration attribute. It applies on type expressions.

I think the solution is to modify `parseParameterTypeExpr` and leave `parseAttributeList` as it is. There's only one "attribute" we expect here (and maybe we'll need `@given` in the future if we change the syntax of implicit parameters).

**Comment by peter-evans** [comment]
> I think the solution is to modify `parseParameterTypeExpr` and leave `parseAttributeList` as it is. There's only one "attribute" we expect here (and maybe we'll need `@given` in the future if we change the syntax of implicit parameters).

Fixed it as you described. https://github.com/peter-evans/hylo/commit/80969a4481e7119ac25a63daee7c10893aeac2da
Just one thing to note, it's now not possible to do this check. If we attempt to parse arguments we'll run into [the issue](https://github.com/hylo-lang/hylo/issues/76#issuecomment-1898820655). Should I remove `.error(attributeTakesNoArgument: a)`, since it's no longer used?
```
        if !a.value.arguments.isEmpty {
          state.diagnostics.insert(.error(attributeTakesNoArgument: a))
        }
```

I still have the other issue I mentioned in the second half of [this comment](https://github.com/hylo-lang/hylo/issues/76#issuecomment-1906275545), because that is caused by attributes now being expressions.

**Comment by kyouko-taiga** [comment]
> If we attempt to parse arguments we'll run into https://github.com/hylo-lang/hylo/issues/76#issuecomment-1898820655.

Yeah, that is unfortunate but I guess we should emit a more precise diagnostic here; something that acknowledges the fact that we're parsing a parameter type annotation. It's fine to leave that task to our future selves, though. Diagnostics in general require _a lot_ of work.

> Should I remove .error(attributeTakesNoArgument: a), since it's no longer used?

Yes, please do!

> I still have the other issue [...]

You can store an expression in the declarations that had the attributes. The "difficult" part will be to extract the value of the expression. The API will probably get a little more complex because the attribute will store an expression ID rather than just a string.

A clean approach would be to validate attributes as soon as we can. For the ones we currently use I think it can be done during parsing. For example, `@foreignName` _must_ have exactly one argument, which must be a simple name expression without any qualification. You could have a look at the way other AST nodes are validated; there's a protocol requirement.

Then I would probably add methods on `AST` to extract arguments from attributes. In the end, things like `ast[d].externalName` will become `ast.externalName(of: d)`, assuming `d` is a `FunctionDecl`. You may also want to extend `BundleNode` for convenience, but that would be a "nice to have".

That's a lot, so don't hesitate to ask if you're lost.

**Comment by peter-evans** [comment]
This is what I have so far.
https://github.com/hylo-lang/hylo/compare/main...peter-evans:hylo:bundled-node-ext

The first commit tries to cover the main goal of the issue. The second commit is an attempt to extend `BundledNode` and add `foreignName`. I'm not sure if this is what you intended the implementation to look like. If it's ok, I'll try to add `externalName`, too.


> A clean approach would be to validate attributes as soon as we can. For the ones we currently use I think it can be done during parsing. For example, @foreignName must have exactly one argument, which must be a simple name expression without any qualification. You could have a look at the way other AST nodes are validated; there's a protocol requirement.

I'm didn't understand this so it's not part of the first commit implementation. If this validation should be done please could you describe in a bit more detail.

**Comment by peter-evans** [comment]
I decided to just open a PR to make it easier to review. https://github.com/hylo-lang/hylo/pull/1336
